api/python/t4/formats.py

```diff
@@ -0,0 +1,418 @@
+            if isinstance(obj, typ):
+                return True
+        return False
+
+    def register(self):
+        """Register this format for automatic usage
+
+        Once registered, a format can be looked up by name, handled object
+        types, and handled filetypes as indicated by extension.
+        """
+        FormatsRegistry.register(self)
+
+    def _update_meta(self, meta, serialization_kwargs={}):
+        if meta is not None:
+            format_meta = meta.get('format', {})
+            format_meta['name'] = self.name
+
+            if serialization_kwargs:
+                format_meta['serialization'] = deepcopy(serialization_kwargs)
+            meta['format'] = format_meta
+
+    def serialize(self, obj, meta=None, **kwargs):
+        """Serialize an object using this format
+
+        Serializes `obj` using this format.  If `meta` is given, it is
+        updated at meta['format']['name'] and (for now) meta['target'].
+
+        If **kwargs are given, they are passed on to the serialization
+        function, and are added to meta['target']['serialization'].
+
+        One of the benefits of this is that once a serialization quirk has
+        been used via **kwargs, the same args are used when updating with
+        the same metadata.
+
+        This is a wrapper for lower-level serializers like `dumps` methods.
+        If overridden and then called via super(), it does nothing.
+
+        Args:
+            obj: object to serialize
+            meta: metadata to update
+            **kwargs: kwargs to send to lower-level serializer. Also included in meta
+        """
+        # deactivated if this method is overridden.
+        if type(self).serialize == Format.serialize:
+            # Method not overridden, so we use the configured one
+            if getattr(self, '_serializer', None) is None:
+                raise NotImplementedError()
+            serialized = self._serializer(obj, **kwargs)
+            self._update_meta(meta, kwargs)
+            return serialized
+
+    def deserialize(self, bytes_obj, meta=None, **kwargs):
+        """Deserialize some bytes using this format
+
+        Converts bytes into an object.
+
+        If `meta['format']['deserialization']` is given, it is passed as
+        kwargs to the lower-level deserializer.
+
+        If **kwargs is given, it overrides metadata kwargs, if present.
+
+        Args:
+            bytes_obj: bytes to deserialize
+            meta: object metadata, may contain deserialization prefs
+            **kwargs: metadata to (potentially) use.
+        """
+        # deactivated if this method is overridden.
+        if type(self).deserialize == Format.deserialize:
+            meta = {} if meta is None else meta
+            if getattr(self, '_deserializer', None) is None:
+                raise NotImplementedError()
+            format_meta = meta.get('format', {})
+            deserialization_kwargs = format_meta.get('deserialization', {})
```

Comment:
I'm not sure this is a good idea... This could potentially allow things like `np.load(..., allow_pickle=True)` (ok, looks like you're not using kwargs for numpy and parquet, but still.) But also, it's very Python-specific.

Reply:
The former -- fine, but the too-python-specific part I can get on board with.  I'll change it.

Reply:
In consideration, even the former is rather an issue -- we shouldn't make it easy for mistakes to be made.